x/tokenfactory/keeper/msg_server.go

```diff
@@ -0,0 +1,169 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/osmosis-labs/osmosis/v7/x/tokenfactory/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (server msgServer) CreateDenom(goCtx context.Context, msg *types.MsgCreateDenom) (*types.MsgCreateDenomResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	denom, err := server.Keeper.CreateDenom(ctx, msg.Sender, msg.Nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.TypeMsgCreateDenom,
+			sdk.NewAttribute(types.AttributeCreator, msg.Sender),
+			sdk.NewAttribute(types.AttributeNewTokenDenom, denom),
+		),
+	})
+
+	return &types.MsgCreateDenomResponse{
+		NewTokenDenom: denom,
+	}, nil
+}
+
+func (server msgServer) Mint(goCtx context.Context, msg *types.MsgMint) (*types.MsgMintResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	authorityMetadata, err := server.Keeper.GetAuthorityMetadata(ctx, msg.Amount.GetDenom())
+	if err != nil {
+		return nil, err
+	}
+
+	if msg.Sender != authorityMetadata.GetAdmin() {
+		return nil, types.ErrUnauthorized
+	}
+
+	// Temporary additional check that only allows sender to mint coins into their own account.
+	// This check will be removed in the future once edge cases can be worked out
+	if msg.Sender != msg.MintToAddress {
+		return nil, sdkerrors.Wrap(types.ErrUnauthorized, "MintToAddress must be the same as Sender address")
+	}
+
+	err = server.Keeper.mintTo(ctx, msg.Amount, msg.MintToAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.TypeMsgMint,
+			sdk.NewAttribute(types.AttributeMintToAddress, msg.MintToAddress),
+			sdk.NewAttribute(types.AttributeAmount, msg.Amount.String()),
+		),
+	})
+
+	return &types.MsgMintResponse{}, nil
+}
+
+func (server msgServer) Burn(goCtx context.Context, msg *types.MsgBurn) (*types.MsgBurnResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	authorityMetadata, err := server.Keeper.GetAuthorityMetadata(ctx, msg.Amount.GetDenom())
+	if err != nil {
+		return nil, err
+	}
+
+	if msg.Sender != authorityMetadata.GetAdmin() {
+		return nil, types.ErrUnauthorized
+	}
+
+	// Temporary additional check that only allows sender to burn coins from their own account.
+	// This check will be removed in the future once edge cases can be worked out
+	if msg.Sender != msg.BurnFromAddress {
+		return nil, sdkerrors.Wrap(types.ErrUnauthorized, "BurnFromAddress must be the same as Sender address")
+	}
```

Comment:
Should we just delete the `BurnFromAddress` from the proto for now, and add it with force transfer in the future? (Its the same problem)

Reply:
I figured we still keep the format the same, instead of adding fields later and making existing integrators wrong.  But I guess we could make it so that after we add the new field, if `BurnFromAddress` is empty, default it to burning from sender?

Should we do that?

Reply:
yeah, I'd prefer we do the latter